src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/SerializationRecord.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Collections.Immutable;
+
+namespace System.Formats.Nrbf;
+
+/// <summary>
+/// Abstract class that represents the serialization record.
+/// </summary>
+/// <remarks>
+///  <para>
+///   Every instance returned to the end user can be either <see cref="PrimitiveTypeRecord{T}"/>,
+///   a <see cref="ClassRecord"/> or an <see cref="ArrayRecord"/>.
+///  </para>
+/// </remarks>
+[DebuggerDisplay("{RecordType}, {Id}")]
+public abstract class SerializationRecord
+{
+    internal SerializationRecord() // others can't derive from this type
+    {
+    }
+
+    /// <summary>
+    /// Gets the type of the record.
+    /// </summary>
+    /// <value>The type of the record.</value>
+    public abstract SerializationRecordType RecordType { get; }
+
+    /// <summary>
+    /// Gets the ID of the record.
+    /// </summary>
+    /// <value>The ID of the record.</value>
+    public abstract SerializationRecordId Id { get; }
+
+    /// <summary>
+    /// Gets the name of the serialized type.
+    /// </summary>
+    /// <value>The name of the serialized type.</value>
+    public abstract TypeName TypeName { get; }
+
+    /// <summary>
+    /// Compares the type name read from the payload against the specified type.
+    /// </summary>
+    /// <remarks>
+    /// <para>This method ignores assembly names.</para>
+    /// <para>This method does NOT take into account member names or their genericTypes.</para>
+    /// </remarks>
+    /// <param name="type">The type to compare against.</param>
+    /// <returns><see langword="true" /> if the serialized type name match provided type; otherwise, <see langword="false" />.</returns>
+    public bool TypeNameMatches(Type type) => Matches(type, TypeName);
+
+    private static bool Matches(Type type, TypeName typeName)
+    {
+        // At first, check the non-allocating properties for mismatch.
+        if (type.IsArray != typeName.IsArray || type.IsConstructedGenericType != typeName.IsConstructedGenericType
+            || type.IsPointer != typeName.IsPointer || type.IsByRef != typeName.IsByRef
+            || type.IsNested != typeName.IsNested
+            || (type.IsArray && type.GetArrayRank() != typeName.GetArrayRank()))
```

Comment:
```suggestion
        // We don't need to check for pointers and references to arrays,
        // as it's impossible to serialize them with BF.
        if (type.IsPointer || type.IsByRef)
            return false;

        // At first, check the non-allocating properties for mismatch.
        if (type.IsArray != typeName.IsArray || type.IsConstructedGenericType != typeName.IsConstructedGenericType
            || type.IsNested != typeName.IsNested
            || (type.IsArray && type.GetArrayRank() != typeName.GetArrayRank()))
```
This check can be simpler